### PR TITLE
Add a 1em bottom margin below the save button

### PR DESCRIPTION
**Pull request by kaapstorm**
Hey @biyeun, I want to add a 1em margin below a button, otherwise it touches a well below it and looks dumb.

But I'm not sure where my class definition should go. I'm sure "utilities.less" is probably the wrong place, but the other places all looked even more wrong.

Also unsure about class naming conventions, if any; "one-em-bottom" would be a bad name if we decided that 10px would be better.

Could you tell me what I should change, or point me to a style guide?

Buddy @millerdev

### CODE REVIEW

**Comment by czue** [comment]
fyi @biyeun is on vacation for two weeks so maybe @dannyroberts or @orangejenny might know?

@kaapstorm style guid is here: https://www.commcarehq.org/styleguide/ but i don't think it answers your question anywhere. 

(this change isn't obviously wrong to me, but I always double check with @biyeun sadly)

File: corehq/apps/style/static/style/lib/bootstrap/less/utilities.less [hunk 1]
```diff
@@ -28,3 +28,8 @@
 .affix {
   position: fixed;
 }
+
+// TODO: Move this to wherever it belongs
+.one-em-bottom {
```
Comment (biyeun) [inline corehq/apps/style/static/style/lib/bootstrap/less/utilities.less]: this isn't the right place. `/static/style/less` is where all the less updates go `lib` should not be touched.

**Comment by benrudolph** [comment]
What about using existing classes we have? it looks like that div has no styles applied to it? not sure exactly what that ko binding adds to the div but maybe try `class="btn btn-primary"`.

**Comment by kaapstorm** [comment]
Thanks @czue. Sorry @biyeun, forgot your vac started now. Enjoy your leave. Thanks @benrudolph I'll check those out.

**Comment by kaapstorm** [comment]
@benrudolph, SaveButton is defined [here][1]. It uses existing Bootstrap button styles, which are defined [here][2]. To add a bottom margin, I think I should add a new class, because I couldn't find a Bootstrap class that does that. I just don't know where to define it -- somewhere under corehq/apps/style/static/style/less -- but where?

... Actually, it seems appstore just puts the style [inline][3]. I'm a little ambivalent about that.

  [1]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/js/main.js#L31
  [2]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/style/static/style/lib/bootstrap/docs/assets/css/bootstrap.css#L3149
  [3]: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/appstore/templates/appstore/partials/sidebar.html#L15

**Comment by dannyroberts** [comment]
Wish I had a stronger opinion, but I'm fine with anything that doesn't edit style/lib like biyeun said

File: corehq/apps/style/static/style/less/components/layouts.less [hunk 2]
```diff
@@ -87,3 +87,35 @@
...
+  margin-top: 10px;
+}
+
+.top20 {
+  margin-top: 20px;
+}
+
+.top1em {
+  margin-top: 1em;
+}
```
Comment (kaapstorm) [inline corehq/apps/style/static/style/less/components/layouts.less]: This feels better to me; it allows other people to use the same classes for layout, it seems like a better place to put these and creates a place to add more if, for some reason, you need a 5px margin, or whatever.

**Comment by orangejenny** [comment]
Apologies for length.

You might want to be applying a style to a parent element, not to the button itself. If there were text next to the button, would it also need a bottom margin so as not to crowd the well? If so, then you should add padding/margin to the `span12` or the `row-fluid` or the containing div, not to the button itself. I'm not sure what the overall page markup looks like here, but seems like it might be a case for styling your container element.

On the other hand (there are so many options!)...this is a partial. Is it always going to be directly above other content? It's possible that instead of styling the button or one of its ancestors, the markup where this template gets included should put some space between usercase_config.html and the content below it. This could be an empty paragraph, could be a blank row, could be a horizontal rule.

In general, styles are better when they're semantic. The rough thing about classes like `top10px` and `top1em` (besides that they're different systems of units) is that they're meaningless. (This is one of the reasons I'm with you on inline styles being sketchy.) One dev will slap on a `top10px` and another will grab the `top20px` and you end up with a site that has a different amounts of "breathing room" in different areas, so it looks just a little bit less than professional but it's hard to put your finger on why. Ideally, we get to a place where we have standard container classes and the person marking up content for a page never has to deal with content bumping up against other content, and they're definitely not deciding between 10px and 20px. At the moment, I'd probably either add a top/bottom margin to the row-fluid (possibly to all row-fluids on the page where this appears), or maybe just put an empty paragraph below the .row-fluid and call it a day.

As an aside, it would be nice if it were more obvious whether to modify /static/style/less or /static/style/lib, I've gotten tripped up on that, too. Maybe a one-line README in both directories?

**Comment by benrudolph** [comment]
:heart: everything you just said

**Comment by czue** [comment]
:+1: especially on the one line README!

File: corehq/apps/locations/tests/__init__.py [hunk 3]
```diff
@@ -1,6 +1,11 @@
 from .test_location_import import *
 from .test_location_types import *
-from .test_locations import *
+from .test_locations import (
```
Comment (dimagimon) [inline corehq/apps/locations/tests/__init__.py]: F401 'LocationGroupTest' imported but unused

Comment (dimagimon) [inline corehq/apps/locations/tests/__init__.py]: F401 'LocationProducts' imported but unused

Comment (dimagimon) [inline corehq/apps/locations/tests/__init__.py]: F401 'LocationsTest' imported but unused

Comment (dimagimon) [inline corehq/apps/locations/tests/__init__.py]: F401 'WrapLocationTest' imported but unused

**Comment by dimagimon** [comment]
There are 3 errors:

* settings.py, line 501 - E265 block comment should start with '# '
* settings.py, line 508 - E265 block comment should start with '# '
* settings.py, line 968 - E265 block comment should start with '# '

**Comment by biyeun** [comment]
everything is static/style/lib is a library and should remain untouched.

**Comment by biyeun** [comment]
I'd like to echo what @orangejenny said, that the margin should be applied to the parent block element, not the button.

**Comment by kaapstorm** [comment]
Thanks @biyeun, I agree. I took @orangejenny's advice, added a short README to static/style/lib, and applied the style at the block level. I also dropped the range of styles I had created for different sized gaps, and replaced them with just two, "top-spacer" and "bottom-spacer", with an explanation for why I made them 9px. It's in [commit 3c3c4bd9](https://github.com/dimagi/commcare-hq/commit/3c3c4bd9d4709bf938968098a54c5c0725831ab8). Please let me know if you think I should change anything.

**Pull request closed without merging**